src/consensus/aft/raft.h

```diff
@@ -1117,6 +1178,61 @@ namespace aft
         append_entries.push_back(std::make_tuple(std::move(ds), i));
       }
 
+      is_execution_pending = true;
+      auto msg = std::make_unique<threading::Tmsg<AsyncExecution>>(
+        execute_append_entries_cb,
+        this,
+        std::move(append_entries),
+        std::move(r),
+        confirm_evidence);
+      if (threading::ThreadMessaging::thread_count > 1)
+      {
+        threading::ThreadMessaging::thread_messaging.add_task(
+          threading::ThreadMessaging::get_execution_thread(
+            threading::MAIN_THREAD_ID),
+          std::move(msg));
+      }
+      else
+      {
+        guard.unlock();
+        msg->cb(std::move(msg));
+      }
+    }
+
+    static void execute_append_entries_cb(
+      std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
+    {
+      msg->data.self->execute_append_entries(
+        msg->data.append_entries, msg->data.r, msg->data.confirm_evidence);
+      msg->cb =
+        reinterpret_cast<void (*)(std::unique_ptr<threading::ThreadMsg>)>(
+          continue_execution);
+      if (threading::ThreadMessaging::thread_count > 1)
+      {
+        threading::ThreadMessaging::thread_messaging.add_task(
+          threading::MAIN_THREAD_ID, std::move(msg));
+      }
+      else
+      {
+        msg->cb(std::move(msg));
+      }
+    }
+
+    static void continue_execution(
+      std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
+    {
+      msg->data.self->is_execution_pending = false;
```

Comment:
Please correct me if I'm wrong but this doesn't seem to be protected by any lock.

Reply:
discussed offline